server/src/main/java/org/elasticsearch/common/settings/AbstractScopedSettings.java

```diff
@@ -757,6 +775,32 @@ private static boolean applyDeletes(Set<String> deletes, Settings.Builder builde
         return null;
     }
 
+    /**
+     * Upgrade all settings eligible for upgrade in the specified settings instance.
+     *
+     * @param settings the settings instance that might contain settings to be upgraded
+     * @return a new settings instance if any settings required upgrade, otherwise the same settings instance as specified
+     */
+    public Settings upgradeSettings(final Settings settings) {
+        final Settings.Builder builder = Settings.builder();
+        boolean changed = false; // track if any settings were upgraded
+        for (final String key : settings.keySet()) {
+            final Setting<?> setting = getRaw(key);
+            final Function<Map.Entry<String, String>, Map.Entry<String, String>> upgrader = settingUpgraders.get(setting);
+            if (upgrader == null) {
+                // the setting does not have an upgrader, copy the setting
+                builder.copy(key, settings);
+            } else {
+                // the setting has an upgrader, so mark that we have changed a setting and apply the upgrade logic
+                changed = true;
+                final Map.Entry<String, String> upgrade = upgrader.apply(new Entry(key, settings));
```

Comment:
Shouldn't the value type be Object, since that is the underlying value type within Settings.Builder?

Reply:
We do not expose the raw objects, instead `get` exposes everything as a `String`. We have the typed methods, but that's for when we get using the underlying `Setting`. We can not strongly type here because we have to have wildcards in the collection of settings upgraders.